bundles/org.dataflowanalysis.analysis.converter/src/org/dataflowanalysis/analysis/converter/ProcessASS.java

```diff
@@ -0,0 +1,219 @@
+			Node previousNode = null;
+			for (AbstractActionSequenceElement<?> ASE : actionSequence.getElements()) {
+				if(ASE instanceof AbstractPCMActionSequenceElement) {
+					previousNode = processActionSequenceElement((AbstractPCMActionSequenceElement)ASE, previousNode);
+				}
+			}
+		}
+	}
+	
+	private Node processActionSequenceElement(AbstractPCMActionSequenceElement pcmASE, Node previousDFDNode) {
+		Node dfdNode = getOrCreateDFDNode(pcmASE);
+		
+		// Add a flow between previous node and current node
+		createFlows(previousDFDNode, dfdNode, pcmASE);
+		
+		return dfdNode;
+	}
+	
+	
+	private void createFlows(Node source, Node dest, AbstractPCMActionSequenceElement pcmASE) {
+		if(source == null || dest == null) {
+			return;
+		}
+		List<DataFlowVariable> FlowVariables = pcmASE.getAllDataFlowVariables();
+		for(DataFlowVariable flowVariable : FlowVariables) {
+			String flowName = flowVariable.variableName();
+			
+			Optional<Flow> optFlow = dfd.getFlows().stream().filter(f -> (
+					f.getSourceNode().equals(source) && 
+					f.getDestinationNode().equals(dest) &&
+					f.getEntityName().equals(flowName))).findFirst();
+			
+			if(optFlow.isPresent()) {
+				return; //possibly modify behavior later on
+			}
+			
+			Flow newFlow = dataflowdiagramFactory.eINSTANCE.createFlow();
+			newFlow.setSourceNode(source);
+			newFlow.setDestinationNode(dest);
+			newFlow.setEntityName(flowName);
+			
+			//Assumption in Palladio: Each flows between two nodes with the same paramenters/that are called the same use the same pin
+			Pin sourceOutPin = findOrCreateOutputPin(source, flowName);
+			Pin destInPin = findOrCreateInputPin(dest, flowName);
+			newFlow.setSourcePin(sourceOutPin);
+			newFlow.setDestinationPin(destInPin);
+			
+			// modify behavior
+			ForwardingAssignment forwarding = datadictionaryFactory.eINSTANCE.createForwardingAssignment();
+			forwarding.setOutputPin(sourceOutPin);
+			source.getBehaviour().getAssignment().add(forwarding);
+			
+			this.dfd.getFlows().add(newFlow);
+		}		
+	}
+	
+	// A pin is equivalent if the same parameters are passed
+	private Pin findOrCreateOutputPin(Node source, String parameters) {
+		Optional<Pin> optPin = source.getBehaviour().getOutPin().stream().filter(p -> p.getEntityName().equals(parameters)).findAny();
+		if(optPin.isPresent()) {
+			return optPin.get();
+		}
+		Pin pin = datadictionaryFactory.eINSTANCE.createPin();
+		pin.setEntityName(parameters);
+		source.getBehaviour().getOutPin().add(pin);
+		return pin;
+	}
+	
+	// I know it is basically a code clone from above but i do not care
+	private Pin findOrCreateInputPin(Node dest, String parameters) {
+		Optional<Pin> optPin = dest.getBehaviour().getInPin().stream().filter(p -> p.getEntityName().equals(parameters)).findAny();
+		if(optPin.isPresent()) {
+			return optPin.get();
+		}
+		Pin pin = datadictionaryFactory.eINSTANCE.createPin();
+		pin.setEntityName(parameters);
+		dest.getBehaviour().getInPin().add(pin);
+		return pin;
+	}
```

Comment:
Code Duplication

Reply:
Almost the whole ass to dfd code is from nicolas

Reply:
As can be seen by my comment above findOrCreateInputPin I wrote this code in quite a hurry. It should still be improved... 
However, in this case, as source/dest nodes and input/output pins are switched in these methods, I think that resolving this minor code duplication could actually make the code less readable/understandable and consequently less maintainable...